### PR TITLE
doc/crimson: document wait states

### PR DESCRIPTION
**Pull request by rzarzynski**
Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test signed

**Comment by tchaikov** [comment]
https://ceph--40282.org.readthedocs.build/en/40282/dev/crimson/wait_states/

File: doc/dev/crimson/wait_states.rst [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+At any moment an instance of ``ClientRequest`` can be only in one of these
+phases. Similarly, an object denoting particular phase can host not more
+than single ``ClientRequest`` the same time. At low-level this is achieved
+through lock a `seastar::shared_mutex` instance which is the heat of any
+of a such phase-denoting object.
+
+As the execution advances,
+request enters next phase and leaves the current one freeing it for another
+``ClientRequest`` instance. All these phases form a pipeline which assures
+the order is preserved.
```
Comment (tchaikov) [inline doc/dev/crimson/wait_states.rst:42]: These phases are chained in the order of execution just like stages in a pipeline. This machinery fulfills the data dependencies required for serving a request.

@rzarzynski i think the order preserving is a more complicated topic. 

1. what is the order, at which scale, order by what?
2. why we need to order. what is the contract?
3. how to we preserve the ordering.

File: doc/dev/crimson/wait_states.rst [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    wait on an object has been recovered
+  ``ClientRequest::PGPipeline::get_obc``
+    wait on an object context becomes locked
+  ``ClientRequest::PGPipeline::process``
+    wait if any other ``MOSDOp`` message is handled against this PG
+
+At any moment an instance of ``ClientRequest`` can be only in one of these
+phases. Similarly, an object denoting particular phase can host not more
+than single ``ClientRequest`` the same time. At low-level this is achieved
+through lock a `seastar::shared_mutex` instance which is the heat of any
```
[the remark this answers is not in the document]
Comment (rzarzynski) [reply doc/dev/crimson/wait_states.rst]: Altered.

File: doc/dev/crimson/wait_states.rst [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  ``ClientRequest::PGPipeline::wait_for_active``
+    wait on a PG becomes ``is_active()``
+  ``ClientRequest::PGPipeline::recover_missing``
+    wait on an object has been recovered
+  ``ClientRequest::PGPipeline::get_obc``
+    wait on an object context becomes locked
+  ``ClientRequest::PGPipeline::process``
+    wait if any other ``MOSDOp`` message is handled against this PG
+
+At any moment an instance of ``ClientRequest`` can be only in one of these
```
[the remark this answers is not in the document]
Comment (rzarzynski) [reply doc/dev/crimson/wait_states.rst]: Altered the doc.

File: doc/dev/crimson/wait_states.rst [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+phases. Similarly, an object denoting particular phase can host not more
+than single ``ClientRequest`` the same time. At low-level this is achieved
```
[the remark this answers is not in the document]
Comment (rzarzynski) [reply doc/dev/crimson/wait_states.rst]: Fixed.

File: doc/dev/crimson/wait_states.rst (hunk 1 above)
Comment (rzarzynski) [reply doc/dev/crimson/wait_states.rst]: > what is the order, at which scale, order by what?
> how to we preserve the ordering.

Extended the paragraph about the `ConnectionPipeline` and `PGPipeline`. Covered the flow where requests from multiple connections target same PG.

> why we need to order. what is the contract?

Added a sentence about keeping the classical OSD's semantic on the beginning of the doc.

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
@rzarzynski could you change the prefix in the title to "doc/crimson: document wait states"?

File: doc/dev/crimson/pipeline.rst [hunk 5]
```diff
@@ -0,0 +1,95 @@
+===========
+The ``ClientRequest`` pipeline
+===========
```
Comment (tchaikov) [inline doc/dev/crimson/pipeline.rst:3]:
```suggestion
==============================
The ``ClientRequest`` pipeline
==============================
```

**Comment by tchaikov** [comment]
@rzarzynski sorry for the confusion, you might need to update the title in the commit message as well.

File: doc/dev/crimson/pipeline.rst (hunk 5 above)
Comment (rzarzynski) [reply doc/dev/crimson/pipeline.rst]: Fixed. Thanks for pointing this out!

**Comment by rzarzynski** [comment]
@tchaikov: no worries, changed in both places.

**Pull request merged**